snapcraft/internal/pluginhandler/_build_attributes.py

```diff
@@ -19,6 +19,9 @@ class BuildAttributes:
     def __init__(self, build_attributes):
         self._attributes = build_attributes
 
+    def enable_patchelf(self):
```

Comment:
I haven't looked at the implications, but can we consolidate to`is_patchelf_enabled()` to get an interface for the single source of truth, applying the default depending on classic vs. strict? Perhaps in PluginHandler?

Reply:
Ah, I now see `_patching_required` below, but perhaps it'd make sense to move that PluginHandler and have pluginhandler just pass the correct bool..?

Reply:
Sure, it's mostly a separation of concerns issue and I didn't want to move that knowledge out of here (it used to be below, I just moved it up). Would you also want the libc bits moved to pluginhandler? Context: https://github.com/snapcore/snapcraft/pull/3277/files#diff-b09ac10e7bee2e923d782b720b7dec51R162

Reply:
How about we drop all changes to _patchelf.py (and my suggestion for `is_patchelf_enabled()`) in favor of an updated conditional here: https://github.com/snapcore/snapcraft/blob/37eb4fd15f04a14f337cf004d02e4e54bd657c4d/snapcraft/internal/pluginhandler/__init__.py#L970

Something like:
```
        if self._build_attributes.no_patchelf() and self._project._snap_meta.confinement == "classic":
            logger.warning(
                "The primed files for part {!r} will not be verified for "
                "correctness or patched: build-attributes: [no-patchelf] "
                "is set.".format(self.name)
            )
        elif  self._project._snap_meta.confinement == "classic" or (self._build_attributes.enable_patchelf() and self._project._snap_meta.confinement != "classic"):
            part_patcher = PartPatcher(
                elf_files=elf_files,
                project=self._project,
                build_attributes=self._build_attributes,
                snap_base_path=self._snap_base_path,
                stage_packages=self._part_properties.get("stage-packages", []),
            )
            part_patcher.patch()
```

Personally I'd prefer to see that PartPatcher is only invoked when patching is required.  To reduce coupling, I'm content to have that done in the PluginHandler.  But perhaps @sergiusens has other thoughts?

Reply:
@cjp256 right, that's fine, but I think that will regress in that it will stop patching in the libc case. I'll look into moving that logic out as well.

Reply:
Alright @cjp256 I pushed a commit that does this. The patcher still needs a significant refactor to completely eliminate coupling, but it seems out of place here. Take a look?

Reply:
Agreed that it's way out of scope here. Thanks for the update, LGTM.  Though I'm not sure why the unit tests are now failing on looking for execstack.  I see you patched it below, though I wonder why that's happening as a result of this.

Reply:
Oh, just because I added some `app` test cases here, which is when execstack is called. Sorry I missed that, I have execstack installed so they passed locally. Should be fixed in the latest commit.